odo/backends/sparksql.py

```diff
@@ -58,19 +91,55 @@ def rdd_to_sqlcontext(ctx, rdd, name=None, dshape=None, **kwargs):
         dshape = dshape.measure
     sql_schema = dshape_to_schema(dshape)
     sdf = ctx.applySchema(rdd, sql_schema)
-    ctx.registerRDDAsTable(sdf, name or next(_names))
+    if name is None:
+        name = next(_names)
+    ctx.registerDataFrameAsTable(sdf, name)
+    ctx.cacheTable(name)
     return sdf
 
 
 @discover.register(SQLContext)
 def discover_sqlcontext(ctx):
-    dshapes = valmap(discover, get_catalog(ctx))
-    return datashape.DataShape(datashape.Record(sorted(dshapes.items())))
+    table_names = list(map(str, ctx.tableNames()))
+    dshapes = sorted(zip(table_names,
+                         map(discover, map(ctx.table, table_names))))
+    return datashape.DataShape(datashape.Record(dshapes))
 
 
-@discover.register(SchemaRDD)
+@discover.register(SparkDataFrame)
 def discover_spark_data_frame(df):
-    return datashape.var * schema_to_dshape(df.schema())
+    return datashape.var * schema_to_dshape(df.schema)
+
+
+def chunk_file(filename, chunksize=1 << 30):
+    """Stream `filename` in chunks of size `chunksize`.
+
+    Parameters
+    ----------
+    filename : str
+        File to chunk
+    chunksize : int, optional, defaults to 1GB
+        Number of bytes to hold in memory at a single time
```

Comment:
Good idea. I shaved off about 45 seconds in some runs by lowering the chunk size:

```python
In [5]: time foo = odo(df, 'jsonlines://foo.json', chunksize=1 << 30)
CPU times: user 186 ms, sys: 4.47 s, total: 4.65 s
Wall time: 2min

In [6]: time foo = odo(df, 'jsonlines://foo.json', chunksize=1 << 27)
CPU times: user 147 ms, sys: 4.76 s, total: 4.9 s
Wall time: 1min 13s

In [7]: time foo = odo(df, 'jsonlines://foo.json', chunksize=1 << 25)
CPU times: user 151 ms, sys: 3.77 s, total: 3.92 s
Wall time: 1min 21s
```